txgossip/txgossip.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// Package txgossip provides a mempool for [Streaming Asynchronous Execution],
+// which is also compatible with AvalancheGo's [gossip] mechanism.
+//
+// [Streaming Asynchronous Execution]: https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/194-streaming-asynchronous-execution
+package txgossip
+
+import (
+	"errors"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p/gossip"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/core"
+	"github.com/ava-labs/libevm/core/txpool"
+	"github.com/ava-labs/libevm/core/types"
+	"github.com/ava-labs/libevm/rlp"
+	"go.uber.org/zap"
+)
+
+var (
+	_ gossip.Gossipable              = Transaction{}
+	_ gossip.Marshaller[Transaction] = Marshaller{}
+	_ gossip.Set[Transaction]        = (*Set)(nil)
+)
+
+// A Transaction is a [gossip.Gossipable] wrapper for a [types.Transaction].
+type Transaction struct {
+	*types.Transaction
+}
+
+// GossipID returns the transaction hash.
+func (tx Transaction) GossipID() ids.ID {
+	return ids.ID(tx.Hash())
+}
+
+// A Marshaller implements [gossip.Marshaller] for [Transaction], based on RLP
+// encoding.
+type Marshaller struct{}
+
+// MarshalGossip returns the [rlp] encoding of the underlying
+// [types.Transaction].
+func (Marshaller) MarshalGossip(tx Transaction) ([]byte, error) {
+	return rlp.EncodeToBytes(tx.Transaction)
+}
+
+// UnmarshalGossip [rlp] decodes the buffer into a [types.Transaction].
+func (Marshaller) UnmarshalGossip(buf []byte) (Transaction, error) {
+	tx := Transaction{new(types.Transaction)}
+	if err := rlp.DecodeBytes(buf, tx.Transaction); err != nil {
+		return Transaction{}, err
+	}
+	return tx, nil
+}
+
+// A Set is a [gossip.Set] wrapping a [txpool.TxPool]. Transactions MAY be added
+// to the pool directly, or via [Set.Add].
+type Set struct {
+	Pool  *txpool.TxPool
+	bloom *gossip.BloomFilter
+}
+
+// NewSet returns a new [gossip.Set] and a cleanup function that MUST be called
+// to release resources.
+func NewSet(logger logging.Logger, pool *txpool.TxPool, bloom *gossip.BloomFilter, targetBloomElements int) (*Set, func() error) {
+	txs := make(chan core.NewTxsEvent)
+	sub := pool.SubscribeTransactions(txs, false)
+	done := make(chan struct{})
+	go maintainBloomFilter(logger, bloom, txs, sub.Err(), done, targetBloomElements)
+
+	return &Set{pool, bloom}, func() error {
+		sub.Unsubscribe()
+		<-done
+		return pool.Close()
```

Comment:
I'm not sure that this warrants changing... But I was a bit surprised that this cleanup function closed the `pool`. I probably would have assumed this it only cleaned up logic specific to the `Set`.

I do feel like a comment on this function is needed at least to convey this information... If it were me, I would probably just expect the pool to be closed by the caller rather than implicitly here.